cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -852,24 +854,232 @@ struct LogbChecked {
+struct RoundImpl<T, RoundMode::TOWARDS_ZERO> {
+  static constexpr enable_if_floating_point<T> Round(const T val) {
+    return std::signbit(val) ? std::ceil(val) : std::floor(val);
+  }
+};
+
+template <typename T>
+struct RoundImpl<T, RoundMode::TOWARDS_INFINITY> {
+  static constexpr enable_if_floating_point<T> Round(const T val) {
+    return std::signbit(val) ? std::floor(val) : std::ceil(val);
+  }
+};
+
+template <typename T>
+struct RoundImpl<T, RoundMode::HALF_DOWN> {
+  static constexpr enable_if_floating_point<T> Round(const T val) {
+    return RoundImpl<T, RoundMode::DOWN>::Round(val);
+  }
+};
+
+template <typename T>
+struct RoundImpl<T, RoundMode::HALF_UP> {
+  static constexpr enable_if_floating_point<T> Round(const T val) {
+    return RoundImpl<T, RoundMode::UP>::Round(val);
+  }
+};
+
+template <typename T>
+struct RoundImpl<T, RoundMode::HALF_TOWARDS_ZERO> {
+  static constexpr enable_if_floating_point<T> Round(const T val) {
+    return RoundImpl<T, RoundMode::TOWARDS_ZERO>::Round(val);
+  }
+};
+
+template <typename T>
+struct RoundImpl<T, RoundMode::HALF_TOWARDS_INFINITY> {
+  static constexpr enable_if_floating_point<T> Round(const T val) {
+    return RoundImpl<T, RoundMode::TOWARDS_INFINITY>::Round(val);
+  }
+};
+
+template <typename T>
+struct RoundImpl<T, RoundMode::HALF_TO_EVEN> {
+  static constexpr enable_if_floating_point<T> Round(const T val) {
+    return std::round(val * T(0.5)) * 2;
+  }
+};
+
+template <typename T>
+struct RoundImpl<T, RoundMode::HALF_TO_ODD> {
+  static constexpr enable_if_floating_point<T> Round(const T val) {
+    return std::floor(val * T(0.5)) + std::ceil(val * T(0.5));
+  }
+};
+
+template <RoundMode RndMode>
+struct Round {
+  using RoundState = OptionsWrapper<RoundOptions>;
+
+  template <typename T, typename Arg>
+  static enable_if_floating_point<Arg, T> Call(KernelContext* ctx, Arg arg, Status* st) {
+    static_assert(std::is_same<T, Arg>::value, "");
+    static_assert(RoundMode::HALF_DOWN > RoundMode::DOWN &&
+                      RoundMode::HALF_DOWN > RoundMode::UP &&
+                      RoundMode::HALF_DOWN > RoundMode::TOWARDS_ZERO &&
+                      RoundMode::HALF_DOWN > RoundMode::TOWARDS_INFINITY &&
+                      RoundMode::HALF_DOWN < RoundMode::HALF_UP &&
+                      RoundMode::HALF_DOWN < RoundMode::HALF_TOWARDS_ZERO &&
+                      RoundMode::HALF_DOWN < RoundMode::HALF_TOWARDS_INFINITY &&
+                      RoundMode::HALF_DOWN < RoundMode::HALF_TO_EVEN &&
+                      RoundMode::HALF_DOWN < RoundMode::HALF_TO_ODD,
+                  "Round modes prefixed with HALF need to be defined last in enum and "
+                  "the first HALF entry has to be HALF_DOWN.");
+
+    auto options = RoundState::Get(ctx);
+    auto pow10 = RoundUtil::Pow10<T>(std::llabs(options.ndigits));
+    if (std::isnan(pow10)) {
+      *st = Status::Invalid("out-of-range value for rounding digits");
+      return arg;
```

Comment:
It would be much better to make the kernel stateful and precompute this in the init function, IMHO.

Reply:
I agree, and I tried to make use of [`KernelStateFromFunctionOptions`](https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/codegen_internal.h#L96) but failed to make it work. The issues I encountered were:
* I do not know how to invoke `Init` from this wrapper (while allowing the existing `Init` to be invoked first). Actually, I calculated `pow10` in the constructor.
* An invalid `status` cannot be handled in constructor, it has to be `Init`.
* `Call` is a static function and member variable `state` is not.

Here is more or less what I tried:
```c++
struct RoundState {
  RoundOptions options_;
  Status status_ = Status::OK();
  double pow10_;

  explicit RoundState(KernelContext* ctx, RoundOptions options) : options_(std::move(options)) {
    pow10_ = RoundUtil::Pow10<double>(std::llabs(options_.ndigits));
    if (std::isnan(pow10_)) {
      // NOTE: There is no way to signal this invalid status outside of Call (e.g., a PreExec method like in StringTransforms).
      status_ = Status::Invalid("out-of-range value for rounding digits");
    }
  }
};

template <RoundMode RndMode>
struct Round {
  using State = KernelStateFromFunctionOptions<RoundState, RoundOptions>;
  const State& state_;

template <typename T, typename Arg>
static enable_if_floating_point<Arg, T> Call(KernelContext* ctx, Arg arg, Status* st) {
    static_assert(std::is_same<T, Arg>::value, "");
    // NOTE: There is no way to signal this invalid status outside of Call (e.g., a PreExec method).
    if (state_.status_ != Status::OK()) {
        *st = state_.status_;
        return arg;
     } else if (!std::isfinite(arg)) {
        return arg;
     }

    auto pow10 = T(state_.pow10_);
    ...
  }
};
```

Reply:
It seems you'll have to use ScalarUnaryNotNullStateful and do some more refactoring on top of that (i.e. you'd be implementing a static Exec which validates the options and instantiates the Op, then calls out to ScalarUnaryNotNullStateful).

Reply:
I was able to solve this issue by creating and specializing `RoundOptionsWrapper` (which is a `KernelState`) to perform:
* validation for `round_to_multiple` in `Init()`
* extends state with `pow10` data member for `round`